### PR TITLE
engine(DM): implement dm openapi

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: ref #5504 
### What is changed and how it works?
- implement openapi for dm
- remove debug job command

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/jobmaster/dm/openapi.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+// DMAPIGetJobConfig implements the api of get job config.
+func (jm *JobMaster) DMAPIGetJobConfig(c *gin.Context) {
+	cfg, err := jm.GetJobCfg(c.Request.Context())
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	bs, err := cfg.Yaml()
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:71]: If the output is directly passed to Web frontend, I guess JSON is more common to be used. Will discuss later

File: engine/jobmaster/dm/openapi.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+// DMAPISetBinlogOperator implements the api of set binlog operator.
+func (jm *JobMaster) DMAPISetBinlogOperator(c *gin.Context, taskName string) {
+	var req openapi.SetBinlogOperatorRequest
+	if err := c.Bind(&req); err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	r := &dmpkg.BinlogRequest{
+		Sources: []string{taskName},
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:141]: seems the field should be `Task`, but in e2e_dm_test.go we use `[]string{source1, source2}` as tasks. Very confusing

File: engine/jobmaster/dm/openapi.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+	}
+	switch req.Op {
+	case openapi.SetBinlogOperatorRequestOpInject:
+		r.Op = pb.ErrorOp_Inject
+	case openapi.SetBinlogOperatorRequestOpSkip:
+		r.Op = pb.ErrorOp_Skip
+	case openapi.SetBinlogOperatorRequestOpReplace:
+		r.Op = pb.ErrorOp_Replace
+	default:
+		_ = c.Error(errors.New("unsupport op type '' for set binlog operator"))
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:157]:
```suggestion
		_ = c.Error(errors.New("unsupported op type '' for set binlog operator"))
```

File: engine/jobmaster/dm/openapi.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+	}
+	if params.BinlogPos != nil {
+		req.BinlogPos = *params.BinlogPos
+	}
+	resp, err := jm.Binlog(c.Request.Context(), req)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	// TODO: use no content response, now return content because delete may failed.
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:182]: Not sure if we should return a 500 errors when resp.ErrorMsg is not empty

File: engine/test/e2e/e2e_dm_test.go [hunk 5]
```diff
@@ -177,154 +185,269 @@ func testSimpleAllModeTask(
...
+
+func queryStatus(ctx context.Context, client *http.Client, jobID string, tasks []string, t *testing.T) (*dm.JobStatus, error) {
+	url := fmt.Sprintf(baseURL+"/status", jobID)
+	for i, task := range tasks {
+		if i == 0 {
+			url += "?"
+		} else {
+			url += "&"
+		}
+		url += "tasks=" + task
```
Comment (lance6716) [inline engine/test/e2e/e2e_dm_test.go:317]: maybe https://pkg.go.dev/net/url#Values.Encode because it also handle special characters encoding

**Review by lance6716 — COMMENTED** [review]
rest lgtm

**Comment by amyangfei** [comment]
PTAL @gozssky @maxshuang

File: engine/jobmaster/dm/openapi.go (hunk 2 above)
Comment (GMHDBJD) [reply engine/jobmaster/dm/openapi.go:141]: In engine we use tasks, in DM we use source...

File: engine/jobmaster/dm/openapi.go [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+	router.PUT("/config", wrapper.DMAPIUpdateJobConfig)
+
+	router.GET("/schema/tasks/:task-name", wrapper.DMAPIGetSchema)
+
+	router.PUT("/schema/tasks/:task-name", wrapper.DMAPISetSchema)
+
+	router.GET("/status", wrapper.DMAPIGetJobStatus)
+
+	router.PUT("/status", wrapper.DMAPIOperateJob)
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:47]: should use POST according to out OpenAPI doc

File: engine/jobmaster/dm/openapi.go [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+	var tasks []string
+	if params.Tasks != nil {
+		tasks = *params.Tasks
+	}
+	resp, err := jm.QueryJobStatus(c.Request.Context(), tasks)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.IndentedJSON(http.StatusOK, resp)
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:61]: should return 500 when status has error

File: engine/jobmaster/dm/openapi.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+	default:
+		_ = c.Error(errors.New("unsupported op type '' for set binlog operator"))
+		return
+	}
+	resp, err := jm.Binlog(c.Request.Context(), r)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.IndentedJSON(http.StatusOK, resp)
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:165]:
```suggestion
	c.IndentedJSON(http.StatusCreated, resp)
```

for POST

File: engine/jobmaster/dm/openapi.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+	if params.BinlogPos != nil {
+		req.BinlogPos = *params.BinlogPos
+	}
+	resp, err := jm.Binlog(c.Request.Context(), req)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	// TODO: use no content response, now return content because delete may failed.
+	c.IndentedJSON(http.StatusOK, resp)
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:183]: for successful deletion, status code is 204

File: engine/jobmaster/dm/openapi.go [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+		return
+	}
+
+	r := &dmpkg.BinlogSchemaRequest{
+		Op:       op,
+		Sources:  []string{taskname},
+		Database: database,
+		Table:    table,
+	}
+	resp := jm.BinlogSchema(c.Request.Context(), r)
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:219]: don't forget to check error message

File: engine/jobmaster/dm/openapi.go (hunk 1 above)
Comment (lance6716) [reply engine/jobmaster/dm/openapi.go:71]: JSON is almost the only accepted format in our doc

File: engine/jobmaster/dm/openapi.go [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+	err := jm.OperateTask(c.Request.Context(), op, nil, tasks)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.Status(http.StatusOK)
+}
+
+// DMAPIGetBinlogOperator implements the api of get binlog operator.
+func (jm *JobMaster) DMAPIGetBinlogOperator(c *gin.Context, taskName string, params openapi.DMAPIGetBinlogOperatorParams) {
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:116]: hope there's not many operators so we don't need to add a `total` field and paging

**Review by lance6716 — COMMENTED** [review]
there's an unified error structure for "API error structure" we can do it in next PR 😂

File: engine/jobmaster/dm/openapi.go (hunk 6 above)
Comment (GMHDBJD) [reply engine/jobmaster/dm/openapi.go:47]: It's not create status, it's change status, so I think PUT is more accurate according to the doc.

File: engine/jobmaster/dm/openapi.go (hunk 11 above)
Comment (GMHDBJD) [reply engine/jobmaster/dm/openapi.go:116]: add a todo

File: engine/jobmaster/dm/openapi.go (hunk 10 above)
Comment (GMHDBJD) [reply engine/jobmaster/dm/openapi.go:219]: add a todo in terrorHTTPErrorHandler

File: engine/jobmaster/dm/openapi.go [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+	switch req.Op {
+	case openapi.SetBinlogOperatorRequestOpInject:
+		r.Op = pb.ErrorOp_Inject
+	case openapi.SetBinlogOperatorRequestOpSkip:
+		r.Op = pb.ErrorOp_Skip
+	case openapi.SetBinlogOperatorRequestOpReplace:
+		r.Op = pb.ErrorOp_Replace
+	default:
+		// nolint:errcheck
+		_ = c.Error(errors.New("unsupported op type '' for set binlog operator"))
```
Comment (amyangfei) [inline engine/jobmaster/dm/openapi.go:192]: Can add `req.Op` in error.

File: tests/utils/many_sorters_test/many_sorters.go [hunk 13]
```diff
@@ -30,6 +30,7 @@ import (
 	"github.com/pingcap/tiflow/cdc/sorter"
 	"github.com/pingcap/tiflow/cdc/sorter/unified"
 	"github.com/pingcap/tiflow/pkg/config"
+	cerrors "github.com/pingcap/tiflow/pkg/errors"
```
Comment (amyangfei) [inline tests/utils/many_sorters_test/many_sorters.go:33]: Is this file(test caes) dead code, it is not reasonable to exist missing import. 🤔

**Review by amyangfei — APPROVED** [review]

File: engine/jobmaster/dm/openapi.go [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+// TODO: extract terror from worker response
+func terrorHTTPErrorHandler() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		c.Next()
+		gErr := c.Errors.Last()
+		if gErr == nil {
+			return
+		}
+		var (
+			code int
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi.go:77]: there should be a default error code, we can do it in next PR. And this function should be put before DMAPIGetJobStatus

File: engine/jobmaster/dm/openapi/spec/dm.yaml [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+          schema:
+            type: string
+      requestBody:
+        description: "request body"
+        content:
+          "application/json":
+            schema:
+              $ref: "#/components/schemas/SetBinlogOperatorRequest"
+      responses:
+        "200":
```
Comment (lance6716) [inline engine/jobmaster/dm/openapi/spec/dm.yaml:216]: not sure what's the effect of HTTP status code in this file. post should be 201

Comment (lance6716) [reply engine/jobmaster/dm/openapi/spec/dm.yaml:216]: rest lgtm

File: engine/jobmaster/dm/openapi.go (hunk 14 above)
Comment (GMHDBJD) [reply engine/jobmaster/dm/openapi.go:78]: add another TODO... Now I think 0 is ok.

**Review by lance6716 — APPROVED** [review]

**Pull request merged**